pkg/pipeline/encode/encode_prom.go

```diff
@@ -261,6 +263,217 @@ func (e *EncodeProm) cleanupExpiredEntriesLoop() {
+		histo: agghistogram,
+		info:  mInfo,
+	}
+}
+
+func (e *EncodeProm) cleanDeletedCounters(newCfg api.PromEncode) {
+	for fullName, counter := range e.counters {
+		if !strings.HasPrefix(fullName, newCfg.Prefix) {
+			e.registerer.Unregister(counter.counter)
+			delete(e.counters, fullName)
+		}
+		metricName := strings.TrimPrefix(fullName, newCfg.Prefix)
+		found := false
+		for _, mCfg := range newCfg.Metrics {
+			if metricName == mCfg.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			e.registerer.Unregister(counter.counter)
+			delete(e.counters, fullName)
+
+		}
+	}
+
+}
+
+func (e *EncodeProm) cleanDeletedGauges(newCfg api.PromEncode) {
+	for fullName, gauge := range e.gauges {
+		if !strings.HasPrefix(fullName, newCfg.Prefix) {
+			e.registerer.Unregister(gauge.gauge)
+			delete(e.gauges, fullName)
+		}
+		metricName := strings.TrimPrefix(fullName, newCfg.Prefix)
+		found := false
+		for _, mCfg := range newCfg.Metrics {
+			if metricName == mCfg.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			e.registerer.Unregister(gauge.gauge)
+			delete(e.gauges, fullName)
+
+		}
+	}
+
+}
+
+func (e *EncodeProm) cleanDeletedHistograms(newCfg api.PromEncode) {
+	for fullName, histogram := range e.histos {
+		if !strings.HasPrefix(fullName, newCfg.Prefix) {
+			e.registerer.Unregister(histogram.histo)
+			delete(e.histos, fullName)
+		}
+		metricName := strings.TrimPrefix(fullName, newCfg.Prefix)
+		found := false
+		for _, mCfg := range newCfg.Metrics {
+			if metricName == mCfg.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			e.registerer.Unregister(histogram.histo)
+			delete(e.histos, fullName)
+
+		}
+	}
+
+}
+
+func (e *EncodeProm) cleanDeletedAggHistograms(newCfg api.PromEncode) {
+	for fullName, aggHisto := range e.aggHistos {
+		if !strings.HasPrefix(fullName, newCfg.Prefix) {
+			e.registerer.Unregister(aggHisto.histo)
+			delete(e.aggHistos, fullName)
+		}
+		metricName := strings.TrimPrefix(fullName, newCfg.Prefix)
+		found := false
+		for _, mCfg := range newCfg.Metrics {
+			if metricName == mCfg.Name {
+				found = true
+				break
+			}
+		}
+		if !found {
+			e.registerer.Unregister(aggHisto.histo)
+			delete(e.aggHistos, fullName)
+
+		}
+	}
+
+}
+
+func (e *EncodeProm) cleandDeletedMetrics(newCfg api.PromEncode) {
+	e.cleanDeletedCounters(newCfg)
+	e.cleanDeletedGauges(newCfg)
+	e.cleanDeletedHistograms(newCfg)
+	e.cleanDeletedAggHistograms(newCfg)
+}
+
+func (e *EncodeProm) checkConfUpdate() {
+	select {
+	case stage := <-e.updateChan:
+		cfg := api.PromEncode{}
+		if stage.Encode != nil && stage.Encode.Prom != nil {
+			cfg = *stage.Encode.Prom
+		}
+
+		e.cleandDeletedMetrics(cfg)
+
+		for _, mCfg := range cfg.Metrics {
+			fullMetricName := cfg.Prefix + mCfg.Name
+			mInfo := CreateMetricInfo(mCfg)
+			switch mCfg.Type {
+			case api.MetricEncodeOperationName("Counter"):
+				if oldMetric, ok := e.counters[fullMetricName]; ok {
+					if !reflect.DeepEqual(mInfo, oldMetric.info) {
```

Comment:
I'm not sure if that works as intended, as `info` contains a predicate (a function) so they'll likely never be equal: https://pkg.go.dev/reflect#DeepEqual
You could test just the api definition `oldMetric.info.MetricsItem`